miniapps/blocksolvers/div_free_solver.cpp

```diff
@@ -0,0 +1,709 @@
+     agg_l2dof_(agg_l2dof), solvers_loc_(agg_l2dof.NumRows())
+{
+   coarse_l2_projector_.Reset(new ProductOperator(&P_l2, &Q_l2, false, false));
+
+   offsets_loc_.SetSize(3, 0);
+   offsets_.SetSize(3, 0);
+   offsets_[1] = M.NumRows();
+   offsets_[2] = M.NumRows() + B.NumRows();
+
+   SparseMatrix M_diag, B_diag;
+   M.GetDiag(M_diag);
+   B.GetDiag(B_diag);
+
+   DenseMatrix B_loc, M_loc;
+
+   for (int agg = 0; agg < solvers_loc_.Size(); agg++)
+   {
+      GetRowColumnsRef(agg_hdivdof_, agg, hdivdofs_loc_);
+      GetRowColumnsRef(agg_l2dof_, agg, l2dofs_loc_);
+      GetSubMatrix(M_diag, hdivdofs_loc_, hdivdofs_loc_, M_loc);
+      GetSubMatrix(B_diag, l2dofs_loc_, hdivdofs_loc_, B_loc);
+      solvers_loc_[agg].Reset(new LocalSolver(M_loc, B_loc));
+   }
+}
+
+void SchwarzSmoother::Mult(const Vector & x, Vector & y) const
+{
+   y.SetSize(offsets_[2]);
+   y = 0.0;
+
+   BlockVector blk_y(y.GetData(), offsets_);
+   BlockVector Pi_x(offsets_); // aggregate-wise average free projection of x
+   static_cast<Vector&>(Pi_x) = x;
+
+   // Right hand side: F_l = F - W_l P_l2[l] (W_{l+1})^{-1} P_l2[l]^T F
+   Vector coarse_l2_projection(Pi_x.BlockSize(1));
+   coarse_l2_projector_->MultTranspose(Pi_x.GetBlock(1), coarse_l2_projection);
+
+   Pi_x.GetBlock(1) -= coarse_l2_projection;
+
+   for (int agg = 0; agg < solvers_loc_.Size(); agg++)
+   {
+      GetRowColumnsRef(agg_hdivdof_, agg, hdivdofs_loc_);
+      GetRowColumnsRef(agg_l2dof_, agg, l2dofs_loc_);
+
+      offsets_loc_[1] = hdivdofs_loc_.Size();
+      offsets_loc_[2] = offsets_loc_[1]+l2dofs_loc_.Size();
+
+      BlockVector rhs_loc(offsets_loc_), sol_loc(offsets_loc_);
+      Pi_x.GetBlock(0).GetSubVector(hdivdofs_loc_, rhs_loc.GetBlock(0));
+      Pi_x.GetBlock(1).GetSubVector(l2dofs_loc_, rhs_loc.GetBlock(1));
+
+      solvers_loc_[agg]->Mult(rhs_loc, sol_loc);
+
+      blk_y.GetBlock(0).AddElementVector(hdivdofs_loc_, sol_loc.GetBlock(0));
+      blk_y.GetBlock(1).AddElementVector(l2dofs_loc_, sol_loc.GetBlock(1));
+   }
+
+   coarse_l2_projector_->Mult(blk_y.GetBlock(1), coarse_l2_projection);
+   blk_y.GetBlock(1) -= coarse_l2_projection;
+}
+
+KernelSmoother::KernelSmoother(const BlockOperator& op,
+                               const HypreParMatrix& kernel_map)
+   : Solver(op.NumRows()), offsets_(2)
+{
+   offsets_[0] = 0;
+   offsets_[1] = kernel_map.NumCols();
+
+   BlockOperator* blk_map = new BlockOperator(op.ColOffsets(), offsets_);
+   blk_map->SetBlock(0, 0, const_cast<HypreParMatrix*>(&kernel_map));
+   blk_kernel_map_.Reset(blk_map);
+
+   const HypreParMatrix& M = dynamic_cast<const HypreParMatrix&>(op.GetBlock(0,
+                                                                             0));
+   kernel_system_.Reset(TwoStepsRAP(kernel_map, M, kernel_map));
+   kernel_system_.As<HypreParMatrix>()->EliminateZeroRows();
+   kernel_system_.As<HypreParMatrix>()->Threshold(1e-14);
+   kernel_smoother_.Reset(new HypreSmoother(*kernel_system_.As<HypreParMatrix>()));
```

Comment:
This makes a default `HypreSmoother` (ie, L1-Gauss-Seidel). Is that always what you want?